test/nerves_hub_link/update_manager_test.exs

```diff
@@ -87,6 +91,51 @@ defmodule NervesHubLink.UpdateManagerTest do
     end
   end
 
+  describe "401 retry" do
+    setup do
+      port = 6543
+      devpath = "/tmp/fwup_output"
+
+      update_payload = %UpdateInfo{
+        firmware_url: "http://localhost:#{port}/test.fw",
+        firmware_meta: %FirmwareMetadata{}
+      }
+
+      {:ok, plug} =
+        start_supervised(
+          {Plug.Cowboy,
+           scheme: :http,
+           plug: {NervesHubLink.Support.HTTPUnauthorizedErrorPlug, report_pid: self()},
+           options: [port: port]}
+        )
+
+      File.rm(devpath)
+
+      {:ok, [plug: plug, update_payload: update_payload, devpath: "/tmp/fwup_output"]}
+    end
+
+    test "retries firmware updates once", %{
+      update_payload: update_payload,
+      devpath: devpath
+    } do
+      fwup_config = %{default_config() | fwup_devpath: devpath}
+
+      {:ok, manager} = UpdateManager.start_link(fwup_config)
+
+      with_mock NervesHubLink.Socket, check_update_available: fn -> update_payload end do
+        assert UpdateManager.apply_update(manager, update_payload, []) == {:updating, 0}
+
+        assert_receive {:fwup, {:progress, 0}}, 1_000
+        assert_receive {:request_error}, 1_000
+        assert_receive {:request_error}, 1_000
+
+        Process.sleep(300)
```

Comment:
Is there any way around having a hard sleep in the test?

Reply:
hmmmm, I added it because the final assert needed time for the `previous_update` status to be set, but I've run some tests over and over without it and it seems to be fine. I've now removed it, and if issues arise then I might need some help.